### ISSUE TITLE
Wishlist for fast tensorflow support and better ux

### ISSUE THREAD

**Issue by MatthewDaggitt** [vehicle-lang/vehicle#623]
@wenkokke's plan.

Separate logical operations for properties and for "slices"
```
# embedding slices into booleans
SliceToBool :: Slice -> Bool

# splitting boolean connectives out into boolean and slice variants
And :: Bool -> Bool -> Bool
AndSlice :: Slice -> Slice -> Slice
Implies :: Bool -> Bool -> Bool
ImpliesSlice :: Slice -> Slice -> Slice
Not :: Bool -> Bool
NotSlice :: Slice -> Slice
Or :: Bool -> Bool -> Bool
OrSlice :: Slice -> Slice -> Slice

# changing the types of comparisons on indexes

EqIndex :: Index -> Index -> Slice
GeIndex :: Index -> Index -> Slice
GtIndex :: Index -> Index -> Slice
LeIndex :: Index -> Index -> Slice
LtIndex :: Index -> Index -> Slice
NeIndex :: Index -> Index -> Slice
```
For the immediate future, we can interpret Slice as a bool that consists only of comparisons on values of type Index, and interpret Bool as Rat, for the loss function semantics, and instantiate it as Never on the Python side to indicate the fact that these should never be returned as part of a loss function. This lets us separate out their interpretation for TensorFlow, which means we get (1) fast graph-based operations, and (2) sensible semantics for indexes.
In the future, we should interpret Slice as numpy index arrays, i.e., Array[int] where int is restricted to 0 and 1.
This interpretation requires a little bit of work to push the logical operations on slices down into the call to indices.

2. Separate application nodes for partial applications of top-level declarations and builtins:
```
App        :: Expr -> [Expr] -> Expr
AppPartial :: Expr -> [Expr] -> Expr
```
For starters, we can say that AppPartial is only ever used with builtins, which would already clean up the generated Python tremendously.
However, the best case scenario is where we change Lam to match App in the exported AST:
```
Lam        :: [Binder] -> Expr -> Expr
```
That would let us bunch together abstractions, which lets us extend the use of AppPartial to applications of top-level declarations (i.e., uses of FreeVar), local declarations (i.e., uses of BoundVar) and anonymous functions (i.e, uses of Lam).
That would let us get rid of all pointless currying. It would let us automatically perform currying whenever an AppPartial is encountered, using the Python partial builtin. Furthermore, it would resolve the user-experience issue, where the user has to figure out the order of arguments for the generated top-level declarations. In the resulting function, the user will just be able to pass the arguments by name, as we indented. (edited)
